packages/block-editor/src/store/actions.js

```diff
@@ -258,14 +302,48 @@ export const moveBlocksUp = createOnMove( 'MOVE_BLOCKS_UP' );
  *
  * @return {Object} Action object.
  */
-export function moveBlockToPosition( clientId, fromRootClientId, toRootClientId, index ) {
-	return {
+export function* moveBlockToPosition( clientId, fromRootClientId, toRootClientId, index ) {
+	const templateLock = yield select(
+		'core/block-editor',
+		'getTemplateLock',
+		fromRootClientId
+	);
+
+	// If locking is equal to all on the original clientId (fromRootClientId),
+	// it is not possible to move the block to any other position.
+	if ( templateLock === 'all' ) {
+		return;
+	}
+
+	const action = {
 		type: 'MOVE_BLOCK_TO_POSITION',
 		fromRootClientId,
 		toRootClientId,
 		clientId,
 		index,
 	};
+	// If moving inside the same root block the move is always possible.
+	if ( fromRootClientId === toRootClientId ) {
+		return action;
+	}
+
+	const blockName = yield select(
+		'core/block-editor',
+		'getBlockName',
+		clientId
+	);
+
+	const canInsertBlock = yield select(
+		'core/block-editor',
+		'canInsertBlockType',
+		blockName,
+		toRootClientId
+	);
+
+	// If moving to other parent block, the move is possible if we can insert a block of the same type inside the new parent block.
+	if ( canInsertBlock ) {
+		return action;
```

Comment:
Should we `yield`, not `return`? Unclear to me what behavior is expected from `redux-routine` for the `return`.